app/src/main/java/ch/epfl/qedit/view/treasurehunt/QuestionLocatorActivity.java

```diff
@@ -0,0 +1,189 @@
+
+    // This is the location service from which we retrieve the phone's location, and the
+    // location manager from which we request permissions
+    private LocationService locService;
+    private PermissionManager permManager;
+
+    // Those are the text views that display the direction and distance to the user
+    private TextView text1;
+    private TextView text2;
+
+    // This button has two usages: allowing the user to answer the question, and allowing him
+    // to grant the location permission
+    private Button button;
+
+    public QuestionLocatorActivity() {}
+
+    @Override
+    /* This method is needed to apply the desired language at the activity startup */
+    protected void attachBaseContext(Context base) {
+        super.attachBaseContext(LocaleHelper.onAttach(base));
+    }
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_question_locator);
+
+        // We retrieve the location from the activity arguments
+        Bundle bundle = Objects.requireNonNull(getIntent().getExtras());
+        questionLoc = (Location) bundle.getParcelable(QUESTION_LOCATION);
+        questionRadius = bundle.getFloat(QUESTION_RADIUS);
+
+        // We retrieve the UI elements and set up the default UI
+        text1 = findViewById(R.id.question_locator_text1);
+        text2 = findViewById(R.id.question_locator_text2);
+        button = findViewById(R.id.question_locator_button);
+        setUnknownUI();
+
+        // We retrieve the two singleton backend services
+        locService = LocServiceFactory.getInstance(getApplicationContext());
+        permManager = PermManagerFactory.getInstance();
+
+        // If we couldn't subscribe, we need to ask for permissions for the location
+        if (!locService.subscribe(this, LOCATION_INTERVAL)) askPermissions();
+    }
+
+    private void askPermissions() {
+        permManager.requestPermissions(this, this, REQUESTED_PERMISSIONS);
+    }
+
+    @Override
+    public void onPermissionResult(String[] permissions, boolean[] granted) {
+        for (int i = 0; i < granted.length; i++) {
+            if (!granted[i]) {
+                // If the permissions were not granted, we just show an error UI to the user
+                // If it is the first time the user denies, we show the button, otherwise we
+                // just give up and show nothing.
+                if (permManager.shouldAskAgain(this, permissions[i])) setPermissionUI();
+                else button.setVisibility(View.GONE);
+
+                return;
+            }
+        }
+
+        // The user finally accepted, so we can hide the button and subscribe to the location
+        locService.subscribe(this, LOCATION_INTERVAL);
+        button.setVisibility(View.GONE);
+    }
+
+    // Sets the text of the UI to inform the user that his position is unknown
+    private void setUnknownUI() {
```

Comment:
Strange name `setUnknownUI`...

Reply:
Yes it is maybe not ideal, but the comment above it clarifies its purpose I think. If you have suggestions, you can tell me ;)